docs/spec/overload.rst

```diff
@@ -99,8 +109,396 @@ variable is not sufficient.
+:term:`materialization <materialize>` of these gradual types could be used
+to disambiguate between two or more of the remaining overloads.
+
+- If none of the arguments evaluate to a gradual type, proceed to step 6.
+- If one or more arguments evaluate to a gradual type but no possible
+  materializations of these types would disambiguate between the remaining
+  overloads, proceed to step 6.
+- If possible materializations of these types would disambiguate between
+  two or more of the remaining overloads and this subset of overloads have
+  consistent return types, proceed to step 6. If the return types include
+  type variables, constraint solving should be applied here before testing
+  for consistency.
+- If none of the above conditions are met, the presence of gradual types
+  leads to an ambiguous overload selection. Assume a return type of ``Any``
+  and stop. This preserves the "gradual guarantee".
+
+
+[Eric's note for reviewers: I'm struggling to come up with an
+understandable and unambiguous way to describe this step.
+Suggestions are welcome.]
+
+[Eric's note for reviewers: Pyright currently does not use return type
+consistency in the above check. Instead, it looks for non-overlapping
+return types. If return types are overlapping (that is, one is a consistent
+subtype of another), it uses the wider return type. Only if there is no
+consistency relationship between return types does it consider it an
+"ambiguous" situation and turns it into an Any. This produces better
+results for users of language servers, but it doesn't strictly preserve
+the gradual guarantee. I'm willing to abandon this in favor of a
+strict consistency check.]
+
+
+Step 6: Choose the first remaining candidate overload as the winning
+match. Evaluate it as if it were a non-overloaded function call and stop.
+
+
+Example 1::
+
+  @overload
+  def example1(x: int, y: str) -> int: ...
+  @overload
+  def example1(x: str) -> str: ...
+
+  example1()  # Error in step 1: no plausible overloads
+  example1(1, "")  # Step 1 eliminates second overload
+  example1("")  # Step 1 eliminates first overload
+
+  example1("", "")  # Error in step 2: no compatible overloads
+  example1(1)  # Error in step 2: no compatible overloads
+
+
+Example 2::
+
+  @overload
+  def example2(x: int, y: str, z: int) -> str: ...
+  @overload
+  def example2(x: int, y: int, z: int) -> int: ...
+
+  def test(val: str | int):
+      # In this example, argument type expansion is
+      # performed on the first two arguments. Expansion
+      # of the third is unnecessary.
+      r1 = example2(1, val, 1)
+      reveal_type(r1)  # Should reveal str | int
+
+      # Here, the types of all three arguments are expanded
+      # without success.
+      example2(val, val, val)  # Error in step 3
+
+
+Example 3::
+
+  @overload
+  def example3(x: int) -> int: ...
+  @overload
+  def example3(x: int, y: int) -> tuple[int, int]: ...
+  @overload
+  def example3(*args: int) -> tuple[int, ...]: ...
```

Comment:
Yes, good catch. The rules should still be well defined in the presence of overlapping overloads (since they appear frequently even in typeshed), but the spec should probably steer clear of them in its own examples. I'll update.